File: tools/romtools/hex-string.js

const isNumber = require('is-number');

function hexString(num, pad, prefix) {
  if (prefix === undefined) prefix = '0x';
  if (num < 0) num = 0xFFFFFFFF + num + 1;
  let hex = num.toString(16).toUpperCase();
  if (isNumber(pad)) {
    hex = hex.padStart(pad, '0');
  } else if (num < 0x0100) {
    hex = hex.padStart(2, '0');
  } else if (num < 0x010000) {
    hex = hex.padStart(4, '0');
  } else if (num < 0x01000000) {
    hex = hex.padStart(6, '0');
  } else {
    hex = hex.padStart(8, '0');
  }
  return (prefix + hex);
}

module.exports = hexString;

File: tools/romtools/hex_string.py
'''
Convert an integer into a string in hexadecimal format. The output string can
optionally be padded with zeros and prefix can optionally be added to the
front of the string.
'''

def hex_string(num, pad=None, prefix='0x'):
    if pad is not None:
        pad = int(pad)
    elif num < 0x0100:
        pad = 2
    elif num < 0x010000:
        pad = 4
    elif num < 0x01000000:
        pad = 6
    else:
        pad = 8

    return '%s%0*X' % (prefix, pad, num)
